govcd/metadata_v2.go

```diff
@@ -848,23 +868,149 @@ func mergeMetadataAndWait(client *Client, requestUri string, metadata map[string
 
 // deleteMetadata deletes metadata associated to the input key from an entity referenced by its URI, then returns the
 // task.
-func deleteMetadata(client *Client, requestUri string, key string, isSystem bool) (Task, error) {
+func deleteMetadata(client *Client, requestUri, name, key string, isSystem bool) (Task, error) {
 	apiEndpoint := urlParseRequestURI(requestUri)
 	if isSystem {
 		apiEndpoint.Path += "/metadata/SYSTEM/" + key
 	} else {
 		apiEndpoint.Path += "/metadata/" + key
 	}
 
+	err := filterMetadataToDelete(client, key, requestUri, name, isSystem, client.IgnoredMetadata)
+	if err != nil {
+		return Task{}, err
+	}
+
 	return client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodDelete, "", "error deleting metadata: %s", nil)
 }
 
 // deleteMetadata deletes metadata associated to the input key from an entity referenced by its URI.
-func deleteMetadataAndWait(client *Client, requestUri string, key string, isSystem bool) error {
-	task, err := deleteMetadata(client, requestUri, key, isSystem)
+func deleteMetadataAndWait(client *Client, requestUri, name, key string, isSystem bool) error {
+	task, err := deleteMetadata(client, requestUri, name, key, isSystem)
 	if err != nil {
 		return err
 	}
 
 	return task.WaitTaskCompletion()
 }
+
+// IgnoredMetadata is a structure that defines the metadata entries that should be ignored by the VCD Client.
+// The filtering works in such a way that all the non-nil pointers in an instance of this struct are evaluated with
+// a logical AND.
+// For example, ignoredMetadata.ObjectType = "org", ignoredMetadata.ObjectName = "foo" will ignore all metadata from
+// Organizations whose name is "foo", with any key and any value.
+// Note: This struct is only used by metadata_v2.go methods.
+// Note 2: Filtering by ObjectName is not possible in the "ByHref" methods from VCDClient.
+type IgnoredMetadata struct {
+	ObjectType *string        // Type of the object that has the metadata as defined in the API documentation https://developer.vmware.com/apis/1601/vmware-cloud-director, for example "catalog", "disk", "org"...
```

Comment:
This is probably a premature question in this sense, but would this pose any problems when working with only OpenAPI based entities (including `getMetadataObjectTypeFromHref`)?

Reply:
Yeah, it would work in a different way. For example in [this PR](https://github.com/vmware/go-vcloud-director/pull/557/files#diff-3e1319d45ed96bc5f0fb18bd53192105cb8238939c6c50c3763aad3b82d319cbR119) maybe we could use the first endpoint path, `entities`. So getMetadataObjectTypeFromHref would need either to be changed or use a variant.

Reply:
ACK. This is a different problem to tackle for future.  Closing it for now